gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -191,8 +198,323 @@ TEST(GaussianMixtureFactor, Error) {
+
+    EXPECT_DOUBLES_EQUAL(
+        sigmoid(mu0, mu1, sigma, sigma, midway - lambda),
+        bn->at(0)->asDiscrete()->operator()(DiscreteValues{{m.first, 1}}),
+        1e-8);
+  }
+  {
+    // Shift by lambda
+    VectorValues given;
+    given.insert(z, Vector1(midway + lambda));
+
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    EXPECT_DOUBLES_EQUAL(
+        sigmoid(mu0, mu1, sigma, sigma, midway + lambda),
+        bn->at(0)->asDiscrete()->operator()(DiscreteValues{{m.first, 1}}),
+        1e-8);
+  }
+}
+
+/* ************************************************************************* */
+/**
+ * Test a simple Gaussian Mixture Model represented as P(m)P(z|m)
+ * where m is a discrete variable and z is a continuous variable.
+ * m is binary and depending on m, we have 2 different means
+ * and covariances each for the
+ * Gaussian distribution around which we sample z.
+ *
+ * The resulting factor graph should eliminate to a Bayes net
+ * which represents a Gaussian-like function
+ * where m1>m0 close to 3.1333.
+ */
+TEST(GaussianMixtureFactor, GaussianMixtureModel2) {
+  using namespace test_gmm;
+
+  double mu0 = 1.0, mu1 = 3.0;
+  double sigma0 = 8.0, sigma1 = 4.0;
+
+  DiscreteKey m(M(0), 2);
+  Key z = Z(0);
+
+  auto hbn = GetGaussianMixtureModel(mu0, mu1, sigma0, sigma1);
+
+  // The result should be a bell curve like function
+  // with m1 > m0 close to 3.1333.
+  // We get 3.1333 by finding the maximum value of the function.
+  VectorValues given;
+  given.insert(z, Vector1(3.133));
+  HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+  HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+  // regression
+  HybridBayesNet expected;
+  expected.emplace_back(
+      new DiscreteConditional(m, "0.325603277954/0.674396722046"));
+
+  EXPECT(assert_equal(expected, *bn));
+}
+
+namespace test_two_state_estimation {
+
+/// Create Two State Bayes Network with measurements
+HybridBayesNet CreateBayesNet(double mu0, double mu1, double sigma0,
+                              double sigma1,
+                              bool add_second_measurement = false,
+                              double prior_sigma = 1e-3,
+                              double measurement_sigma = 3.0) {
+  DiscreteKey m1(M(1), 2);
+  Key z0 = Z(0), z1 = Z(1), f01 = F(0);
+  Key x0 = X(0), x1 = X(1);
+
+  HybridBayesNet hbn;
+
+  auto prior_model = noiseModel::Isotropic::Sigma(1, prior_sigma);
+  auto measurement_model = noiseModel::Isotropic::Sigma(1, measurement_sigma);
+
+  // Set a prior P(x0) at x0=0
+  hbn.emplace_back(
+      new GaussianConditional(x0, Vector1(0.0), I_1x1, prior_model));
+
+  // Add measurement P(z0 | x0)
+  auto p_z0 = new GaussianConditional(z0, Vector1(0.0), -I_1x1, x0, I_1x1,
+                                      measurement_model);
+  hbn.emplace_back(p_z0);
+
+  // Add hybrid motion model
```

Comment:
I think this is still wrong :-/
I think we simply need 
```
P(x1| x0,m) = 
N(x0+mu0,sigma0) if m=0
N(x1+mu1, sigma1) if m=1
```
There is no need for a dummy measurement f01.

Reply:
Yes, $x0+mu1$, sorry.
You should not need a measurement at all. $P(x1|x0,x1)$ becomes a factor $\phi(x0,x1,m)$. If our code can’t handle this we learned something :-) let’s chat.

Reply:
Yup, the code doesn't handle this, since to convert to a likelihood, we need to provide the frontal variable. The likelihood then removes the frontal variable, which means $P(x1|x0, m1)$ becomes $\phi(x0, m1)$.

Reply:
As we discussed, I think it does. Eager to see it work :-)